### PR TITLE
Create new SA token for each cluster joined

### PR DESCRIPTION
**Pull request by Jaanki**
Create a new SA/Role/role binding for each cluster
that is joined to the broker.

Solves issue: #100
Signed-By: Janki Chhatbar <jchhatba@redhat.com>

### CODE REVIEW

File: pkg/subctl/cmd/join.go [hunk 1]
```diff
@@ -301,5 +307,14 @@ func populateSubmarinerSpec(subctlData *datafile.SubctlData) submariner.Submarin
 		Namespace:                SubmarinerNamespace,
 	}
 
+	print("submariner spec is %s", submarinerSpec)
 	return submarinerSpec
 }
+
+func createSAPerCluster(clientSet clientset.Interface, brokerNamespace string) (*v1.Secret) {
+    broker.NewBrokerSA()
+    broker.NewBrokerRole()
```
Comment (mangelajo) [inline pkg/subctl/cmd/join.go]: You don't need to create a new role every time, we can have just one role that will be attached to each new SA created for clusters. (via rolebinding)

File: pkg/subctl/operator/install/embeddedyamls/yamls.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
+// This file is auto-generated by yamls2go.go
```
Comment (mangelajo) [inline pkg/subctl/operator/install/embeddedyamls/yamls.go]: Please don't add this file, it's auto generated by our makefile.

File: pkg/subctl/cmd/join.go [hunk 3]
```diff
@@ -301,5 +307,14 @@ func populateSubmarinerSpec(subctlData *datafile.SubctlData) submariner.Submarin
 		Namespace:                SubmarinerNamespace,
 	}
 
+	print("submariner spec is %s", submarinerSpec)
 	return submarinerSpec
 }
+
+func createSAPerCluster(clientSet clientset.Interface, brokerNamespace string) (*v1.Secret) {
+    broker.NewBrokerSA()
```
Comment (mangelajo) [inline pkg/subctl/cmd/join.go]: You didn't add those methods in the PR

Comment (Jaanki) [reply pkg/subctl/cmd/join.go]: I did at L194

File: pkg/subctl/cmd/join.go [hunk 4]
```diff
@@ -441,6 +453,17 @@ func populateSubmarinerSpec(subctlData *datafile.SubctlData) submariner.Submarin
...
 		submarinerSpec.GlobalCIDR = globalCIDR
 	}
-
+	fmt.Printf("submariner spec is %s", submarinerSpec)
 	return submarinerSpec
 }
+
+func createSAPerCluster(clientset *kubernetes.Clientset, brokernamespace string) (*v1.Secret) {
+	broker.CreateNewBrokerSA(clientset, clusterID+"-submariner-k8s-broker-client")
+	serviceaccount, _ := clientset.CoreV1().ServiceAccounts(broker.SubmarinerBrokerNamespace).List(metav1.ListOptions{})
```
Comment (skitt) [inline pkg/subctl/cmd/join.go]: This ignores any errors, which might give you some hint about why this is failing.

File: pkg/subctl/cmd/join.go [hunk 5]
```diff
@@ -441,6 +453,17 @@ func populateSubmarinerSpec(subctlData *datafile.SubctlData) submariner.Submarin
...
+	fmt.Printf("serviceaccount is %s", serviceaccount)
+	broker.CreateNewBrokerRole(clientset, clusterID+"-submariner-k8s-broker-client")
+	broker.CreateNewBrokerRoleBinding(clientset, clusterID+"-submariner-k8s-broker-client", clusterID+"-submariner-k8s-broker-client")
+	broker.WaitForClientToken(clientset)
+	clienttoken, _ = broker.GetClientTokenSecret(clientset, brokernamespace, clusterID+"-submariner-k8s-broker-client")
```
Comment (skitt) [inline pkg/subctl/cmd/join.go]: Same here, if an error occurs it would be helpful to log it.

File: pkg/subctl/cmd/join.go [hunk 6]
```diff
@@ -441,6 +450,19 @@ func populateSubmarinerSpec(subctlData *datafile.SubctlData) submariner.Submarin
...
 		submarinerSpec.GlobalCIDR = globalCIDR
 	}
-
 	return submarinerSpec
 }
+
+func createSAPerCluster(clientset *kubernetes.Clientset, brokernamespace string) *v1.Secret {
+	newsa, err := broker.CreateNewBrokerSA(clientset, clusterID+"-submariner-k8s-broker-client")
+	exitOnError("Service account not created", err)
+	newrole, err := broker.CreateNewBrokerRole(clientset, newsa.Name)
```
Comment (mangelajo) [inline pkg/subctl/cmd/join.go:459]: We don't need to keep a new role for every new cluster. You can have a single role, and then using role bindings to associate the SA to the Role.

Also in this case you'd need to use the explained above "NewClusterBrokerRole", to ensure at least that common role is created.

Comment (Jaanki) [reply pkg/subctl/cmd/join.go:459]: I tried to not create a new role for every cluster but instead have all clusters bind to the same role. This somehow didn't work. I think we need to group SAs and then bind the role to that group. Searching for syntax to do this.

File: pkg/broker/rbac.go [hunk 7]
```diff
@@ -41,7 +36,7 @@ func NewBrokerSA() *v1.ServiceAccount {
 	return sa
 }
 
-func NewBrokerRole() *rbacv1.Role {
+func NewBrokerRole(submarinerBrokerRole string) *rbacv1.Role {
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/broker/rbac.go:39]: Agree. That's a better approach. I thought the same thoughts about giving access to all clusters but didn't think to separate roles.

File: pkg/subctl/cmd/join.go (hunk 6 above)
Comment (mangelajo) [reply pkg/subctl/cmd/join.go:459]: I mean something like this: 
```
[majopela@bluehat submariner]$ kind create cluster
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.15.3) 🖼 
 ✓ Preparing nodes 📦 
 ✓ Creating kubeadm config 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Cluster creation complete. You can now use the cluster with:
export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"
kubectl cluster-info
[majopela@bluehat submariner]$ export KUBECONFIG="$(kind get kubeconfig-path --name="kind")"
[majopela@bluehat submariner]$ kubectl create role pod-reader --verb=get --verb=list --verb=watch --resource=pods
role.rbac.authorization.k8s.io/pod-reader created
[majopela@bluehat submariner]$ kubectl create sa sa-1
serviceaccount/sa-1 created
[majopela@bluehat submariner]$ kubectl create sa sa-2
serviceaccount/sa-2 created
[majopela@bluehat submariner]$ kubectl create rolebinding sa-1-pod-reader --role=pod-reader --serviceaccount=default:sa-1
rolebinding.rbac.authorization.k8s.io/sa-1-pod-reader created
[majopela@bluehat submariner]$ kubectl create rolebinding sa-2-pod-reader --role=pod-reader --serviceaccount=default:sa-2
rolebinding.rbac.authorization.k8s.io/sa-2-pod-reader created
```

File: pkg/broker/rbac.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/broker/rbac.go:39]: I did 

// Create a role to bind to cluster specific SA
func NewSubctlBrokerRole(submarinerSubctlRole string) *rbacv1.Role {
	subctlrole := &rbacv1.Role{
		ObjectMeta: metav1.ObjectMeta{
			Name: submarinerSubctlRole,
		},
		Rules: []rbacv1.PolicyRule{
			rbacv1.PolicyRule{
				APIGroups: []string{"submariner.io"},
				Resources: []string{"clusters", "endpoints"},
				Verbs:     []string{"create", "get", "list", "watch", "patch", "update", "delete"},
			},
			rbacv1.PolicyRule{
				Verbs:     []string{"create", "get", "list", "delete"},
				APIGroups: []string{""},
				Resources: []string{"serviceaccounts", "secrets"},
			},
			rbacv1.PolicyRule{
				Verbs:     []string{"create", "get", "list", "delete"},
				APIGroups: []string{"rbac.authorization.k8s.io"},
				Resources: []string{"roles", "rolebindings"},
			},
		},
	}

	return subctlrole
}

// Create a role for Broker SA to bind to
func NewClusterBrokerRole() *rbacv1.Role {
	clusterbrokerrole := &rbacv1.Role{
		ObjectMeta: metav1.ObjectMeta{
			Name: submarinerBrokerRole,
		},
		Rules: []rbacv1.PolicyRule{
			rbacv1.PolicyRule{
				APIGroups: []string{"submariner.io"},
				Resources: []string{"clusters", "endpoints"},
				Verbs:     []string{"create", "get", "list", "watch", "patch", "update", "delete"},
			},
			rbacv1.PolicyRule{
				Verbs:     []string{"create", "delete"},
				APIGroups: []string{""},
				Resources: []string{"serviceaccounts"},
			},
			rbacv1.PolicyRule{
				Verbs:     []string{"create", "delete"},
				APIGroups: []string{"rbac.authorization.k8s.io"},
				Resources: []string{"roles", "rolebindings"},
			},
		},
	}

	return clusterbrokerrole
}

and created a new rolebinding "subctlrole". On joining the clusters, error is
Role binding failed: rolebindings.rbac.authorization.k8s.io "subctlrole" is forbidden: user "system:serviceaccount:submariner-k8s-broker:submariner-k8s-broker-client" (groups=["system:serviceaccounts" "system:serviceaccounts:submariner-k8s-broker" "system:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:[""], Resources:["secrets"], Verbs:["create" "get" "list" "delete"]}
{APIGroups:[""], Resources:["serviceaccounts"], Verbs:["get" "list"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["rolebindings"], Verbs:["get" "list"]}
{APIGroups:["rbac.authorization.k8s.io"], Resources:["roles"], Verbs:["get" "list"]}

This basically means that the original SA can give only those rights to the new SA that it has. So the concept of 2 roles might not work.

**Pull request merged**